### PR TITLE
Basic setup and Index routes

### PR DESCRIPTION
**Pull request by thapar-vansh**

### CODE REVIEW

File: routes/indexRoutes.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import express from 'express'
+import { register, login, getPlayers } from '../controllers/indexController'
+import { verifyUser } from '../middleware/auth'
+
+export const router = express.Router()
```
Comment (jagdishcm-sixer) [inline routes/indexRoutes.ts]: Type ??

File: controllers/adminController.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  getPlayerById,
+  deletePlayer,
+} from '../services/adminService'
+
+export const addPlayers = async (
+  req: Request,
+  res: Response
+): Promise<Response> => {
+  const { name, country } = req.body
+  if (!name || !country) {
```
Comment (jagdishcm-sixer) [inline controllers/adminController.ts]: should be && instead of ||

File: controllers/adminController.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    return res.status(400).send('Something went wrong')
+  }
+}
+
+export const updatePlayers = async (
+  req: Request,
+  res: Response
+): Promise<Response> => {
+  const { id, name, country } = req.body
+  if (!id || !name || !country) {
```
Comment (jagdishcm-sixer) [inline controllers/adminController.ts]: id should be mandatory. name and country can be optional

File: controllers/indexController.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        )
+    }
+    console.log(e)
+    return res.status(400).send('Something went wrong')
+  }
+}
+
+export const login = async (req: Request, res: Response): Promise<Response> => {
+  const { username, password } = req.body
+  if (!username || !password) {
```
Comment (jagdishcm-sixer) [inline controllers/indexController.ts]: should be && instead of ||

File: controllers/indexController.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    return res.status(400).send('Something went wrong')
+  }
+}
+
+export const register = async (
+  req: Request,
+  res: Response
+): Promise<Response> => {
+  const { username, password } = req.body
+  if (!username || !password) {
```
Comment (jagdishcm-sixer) [inline controllers/indexController.ts]: should be && instead of ||

File: database/players.db.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+export const getPlayerByIdFromDb = (id: number): Promise<QueryResult> =>
+  query(
+    `SELECT * FROM players
+        WHERE id = $1`,
+    [id]
+  ) as Promise<QueryResult>
+
+export const getPlayersFromDb = (): Promise<QueryResult> =>
+  query(`SELECT * FROM players`, []) as Promise<QueryResult>
+// const players: string[] = []
```
Comment (jagdishcm-sixer) [inline database/players.db.ts]: remove all the unneeded commented code before creating a PR

File: middleware/auth.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  const token: string = req.headers['admintoken'] as string
+  if (!token) {
+    return res.status(403).send('Token is required for authentication')
+  }
+  try {
+    const decoded = jwt.verify(token, config.JWT_PRIVATEKEY)
```
Comment (jagdishcm-sixer) [inline middleware/auth.ts]: Type ?

File: middleware/auth.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    req.userToken = decoded as (string | number)[]
+    const user = await getUserByUserId(req.userToken['userId'])
```
Comment (jagdishcm-sixer) [inline middleware/auth.ts]: Type ?

File: routes/adminRoutes.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import express from 'express'
+
+import {
+  addPlayers,
+  updatePlayers,
+  deletePlayers,
+} from '../controllers/adminController'
+import { verifyAdmin } from '../middleware/auth.js'
+
+export const router = express.Router()
```
Comment (jagdishcm-sixer) [inline routes/adminRoutes.ts]: Type ?

**Review by jagdishcm-sixer — CHANGES_REQUESTED** [review]

File: tests/services/adminService.test.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    })
+
+    it('service to add player to database', async () => {
+        const mockAddPlayer = jest
+          .spyOn(admin, 'addPlayer')
+          .mockImplementation(() => Promise.resolve(data.insertSuccess))
+        const result = await addPlayer('vansh', 'india')
+        expect(mockAddPlayer).toHaveBeenCalledTimes(1)
+        expect(result).toEqual(data.insertSuccess)
+      })
```
Comment (jagdishcm-sixer) [inline tests/services/adminService.test.ts]: while committing code, check if code has been formatted correctly
PS : checkout prettier code formatter

File: tests/services/adminService.test.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    
+      it('service to get player by id', async () => {
+        const mockGetPlayerById = jest
+          .fn()
+          .mockImplementation(() => Promise.resolve(getPlayerById(1)))
+        const result = await mockGetPlayerById(1)
+        const expectedPlayer = { id: '1', name: 'virat', country: 'india' }
+        expect(mockGetPlayerById).toBeCalledTimes(1)
+        expect(result).toEqual(expectedPlayer)
+      })
```
Comment (jagdishcm-sixer) [inline tests/services/adminService.test.ts]: add negative tests as well 
e.g. when user tries to fetch player with invalid name

File: tests/services/loginService.test.ts [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+  afterEach(() => {
+    jest.clearAllMocks()
+    jest.resetAllMocks()
+  })
+  it('returns all players', async () => {
+    const mockGetAllPlayers = jest
+      .fn()
+      .mockImplementation(() => Promise.resolve(getAllPlayers()))
+
+    const expectedUsers = [
```
Comment (jagdishcm-sixer) [inline tests/services/loginService.test.ts]: put these expected values in a different file to keep the code clean

File: tests/services/loginService.test.ts [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    ]
+
+    expect(await mockGetAllPlayers()).toEqual(expectedUsers)
+  })
+
+  it('registers user', async () => {
+    
+    const mockRegisterUser = jest
+      .fn()
+      .mockImplementation(() => Promise.resolve(registerUser('vansh','India@07')))
```
Comment (jagdishcm-sixer) [inline tests/services/loginService.test.ts]: it's not the correct way of mocking a function for unit test

File: tests/services/userService.test.ts [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+  it('service to check duplicate favourite player from database', async () => {
+    let playerId:number 
+    let userId :number
+
+    const mockCheckDuplicateFav = jest
+      .fn()
+      .mockImplementation(() =>
+        Promise.resolve(checkDuplicateFav(playerId, userId))
+      )
+      //console.log(await checkDuplicateFav(1,20),await checkDuplicateFav(1,2))
```
Comment (jagdishcm-sixer) [inline tests/services/userService.test.ts]: remove the comments and logs too

File: tests/services/userService.test.ts [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+    const result1 = await mockCheckDuplicateFav(1, 20)
+    console.log(result1)
```
Comment (jagdishcm-sixer) [inline tests/services/userService.test.ts]: remove logs

File: tests/services/userService.test.ts [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    expect(result1).toEqual(true)
+
+    const result2 = await mockCheckDuplicateFav(1, 2)
+    expect(result2).toEqual(null)
+
+    console.log(result1,result2)
```
Comment (jagdishcm-sixer) [inline tests/services/userService.test.ts]: remove logs

File: services/loginService.ts [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+      return token
+    } else {
+      throw new Error('Invalid credentials')
+    }
+  } else {
+    throw new Error('User does not exists')
+  }
+}
+
+export function generateToken(userId: number): string {
```
Comment (Sachin-R-Yawalkar) [inline services/loginService.ts]: Change this to es6 function

**Review by Sachin-R-Yawalkar — CHANGES_REQUESTED** [review]

**Review by jagdishcm-sixer — CHANGES_REQUESTED** [review]

File: tests/services/loginService.test.ts [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+    expect(mockGetAllPlayers).toBeCalledTimes(1)
+  })
+
+  it('register user', async () => {
+    const mockRegisterUser = jest
+      .spyOn(userDb, 'getUserbyUsername')
+      .mockResolvedValue(data.registerUserInsertSuccess)
+    const result = await registerUser('testj', 'India@07')
+    expect(mockRegisterUser).toBeCalledTimes(1)
+    expect(result).toBe(undefined)
```
Comment (Sachin-R-Yawalkar) [inline tests/services/loginService.test.ts:37]: action: need to mock this function as well `await addUserToDb(username, hashedPassword)`

File: tests/services/loginService.test.ts [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+  it('returns all players', async () => {
+    const mockGetAllPlayers = jest
+      .spyOn(playerDb, 'getPlayersFromDb')
+      .mockResolvedValueOnce(data.getAllPlayersSuccess)
+    const result = getAllPlayers()
+    expect(await result).toBe(data.getAllPlayersSuccess.rows)
+    expect(mockGetAllPlayers).toBeCalledTimes(1)
+  })
+
+  it('register user', async () => {
```
Comment (jagdishcm-sixer) [inline tests/services/loginService.test.ts]: also specify if test is getting successful or failed

File: tests/services/loginService.test.ts [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+    const mockLoginUser = jest
+      .spyOn(userDb, 'getUserbyUsername')
+      .mockResolvedValueOnce(data.getUserByUsernameFailure)
+
+    const result = loginUser('vansht', 'India@07')
+    expect(await result).toThrowError('User does not exists')
+    expect(mockLoginUser).toBeCalledTimes(1)
+  })
+  it('generates token when user exists', async () => {
+    const mockGenerateToken = jest
```
Comment (jagdishcm-sixer) [inline tests/services/loginService.test.ts:67]: you need to mock getUserbyUsername function as well

**Review by jagdishcm-sixer — CHANGES_REQUESTED** [review]

File: tests/services/userService.test.ts [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+  it('service to delete favourite player from database', async () => {
+    const mockDeleteFavPlayer = jest
+      .spyOn(user, 'deleteFavPlayer')
+      .mockResolvedValue(Promise.resolve(data.deleteSuccess))
+    const result = await deleteFavPlayer(20, 1)
+    expect(mockDeleteFavPlayer).toBeCalledTimes(1)
+    expect(result).toEqual(data.deleteSuccess)
+  })
+  it('service to check duplicate favourite player returns true if player exists ', async () => {
+    const mockCheckDuplicateFav = jest
```
Comment (Sachin-R-Yawalkar) [inline tests/services/userService.test.ts]: action: need to mock `await getPlayerById(id)` before

File: tests/services/userService.test.ts [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+      .spyOn(favourite, 'checkDuplicateFavFromDb')
+      .mockResolvedValueOnce(
+        Promise.resolve(data.checkDuplicateFavSuccessFromDb)
+      )
+    const result = checkDuplicateFav(1, 20)
+    expect(await result).toBe(true)
+    expect(mockCheckDuplicateFav).toBeCalledTimes(1)
+  })
+  it('service to check duplicate favourite player returns null if duplicate does not exist', async () => {
+    const mockCheckDuplicateFav = jest
```
Comment (Sachin-R-Yawalkar) [inline tests/services/userService.test.ts:73]: same as above

**Review by Sachin-R-Yawalkar — CHANGES_REQUESTED** [review]

**Review by jagdishcm-sixer — APPROVED** [review]

**Review by Sachin-R-Yawalkar — APPROVED** [review]
Good Work! Approved

**Pull request merged**